mm2src/coins/eth/eth_balance_events.rs

```diff
@@ -1,32 +1,61 @@
+use super::EthCoin;
+use crate::{eth::{u256_to_big_decimal, Erc20TokenInfo},
+            BalanceError, CoinWithDerivationMethod};
+use common::{executor::Timer, log, Future01CompatExt};
+use mm2_event_stream::{Broadcaster, Event, EventStreamer, NoDataIn, StreamHandlerInput};
+use mm2_number::BigDecimal;
+
 use async_trait::async_trait;
-use common::{executor::{AbortSettings, SpawnAbortable, Timer},
-             log, Future01CompatExt};
 use ethereum_types::Address;
-use futures::{channel::oneshot::{self, Receiver, Sender},
-              stream::FuturesUnordered,
-              StreamExt};
+use futures::{channel::oneshot, stream::FuturesUnordered, StreamExt};
 use instant::Instant;
-use mm2_core::mm_ctx::MmArc;
-use mm2_err_handle::prelude::MmError;
-use mm2_event_stream::{behaviour::{EventBehaviour, EventInitStatus},
-                       ErrorEventName, Event, EventName, EventStreamConfiguration};
-use mm2_number::BigDecimal;
+use serde::Deserialize;
+use serde_json::Value as Json;
 use std::collections::{HashMap, HashSet};
 
-use super::EthCoin;
-use crate::{eth::{u256_to_big_decimal, Erc20TokenInfo},
-            BalanceError, CoinWithDerivationMethod, MmCoin};
+#[derive(Deserialize)]
+#[serde(deny_unknown_fields, default)]
+struct EthBalanceStreamingConfig {
+    /// The time in seconds to wait before re-polling the balance and streaming.
+    pub stream_interval_seconds: f64,
+}
+
+impl Default for EthBalanceStreamingConfig {
+    fn default() -> Self {
+        Self {
+            stream_interval_seconds: 10.0,
+        }
+    }
+}
+
+pub struct EthBalanceEventStreamer {
+    /// The period in seconds between each balance check.
+    interval: f64,
+    coin: EthCoin,
+}
+
+impl EthBalanceEventStreamer {
+    pub fn try_new(config: Option<Json>, coin: EthCoin) -> serde_json::Result<Self> {
+        let config: EthBalanceStreamingConfig = config.map(serde_json::from_value).unwrap_or(Ok(Default::default()))?;
+
+        Ok(Self {
+            interval: config.stream_interval_seconds,
+            coin,
+        })
+    }
+}
```

Comment:
this constructor doesn't need to be `try_new` but `new` which should return `Self` instead. I mean we have default in place so all failure should lead to using default value.

does that make sense ?

Reply:
nope, the default kicks in if the config is `None`, but if the config is invalid we still fail.
looking at this though im asking myself why do that instead of letting serde do the work of failing on a bad deserializaiton. i guess the only plausable answer is that i might have just wanted the config structs to be private.

`pub`ing them and converting things to `new()` instead of `try_new()` though, as this indeed isn't good looking.